control-plane/pkg/reconciler/trigger/trigger.go

```diff
@@ -117,7 +119,7 @@ func (r *Reconciler) reconcileKind(ctx context.Context, trigger *eventing.Trigge
 	}
 
 	// Get data plane config map.
-	contractConfigMap, err := r.GetOrCreateDataPlaneConfigMap(ctx)
+	contractConfigMap, err := r.GetOrCreateDataPlaneConfigMap(ctx, r.DataPlaneConfigMapLabeler)
```

Comment:
This additional parameter should not be necessary.

Reply:
Did not get what you mean...

The `GetOrCreateDataPlaneConfigMap` is defined in the `base.Reconciler` and is used by various reconciler implementations. In some we pass something, in another we pass nothing.

Can you elaborate?

Reply:
This is the idea to remove duplicated code that is in `reconcileDataPlaneConfigMap` (this is bad also because is based on how current the existing reconciler works internally) and also to have `GetOrCreateDataPlaneConfigMap` to have a reconciler semantic:

```
pierdipi@pierdipi eventing-kafka-broker (2022-06-27-namespaced-broker-impl) $ cat remove_duplicate_code.patch 
diff --git a/control-plane/pkg/reconciler/base/reconciler.go b/control-plane/pkg/reconciler/base/reconciler.go
index 1565f6dd..62b05fbd 100644
--- a/control-plane/pkg/reconciler/base/reconciler.go
+++ b/control-plane/pkg/reconciler/base/reconciler.go
@@ -85,6 +85,8 @@ type Reconciler struct {
 
        DispatcherLabel string
        ReceiverLabel   string
+
+       DataPlaneConfigMapTransformer ConfigMapOption
 }
 
 func (r *Reconciler) IsReceiverRunning() bool {
@@ -110,36 +112,36 @@ type ConfigMapOption func(cm *corev1.ConfigMap)
 
 func NoopConfigmapOption(cm *corev1.ConfigMap) {}
 
-func (r *Reconciler) GetOrCreateDataPlaneConfigMap(ctx context.Context, options ...ConfigMapOption) (*corev1.ConfigMap, error) {
+func (r *Reconciler) GetOrCreateDataPlaneConfigMap(ctx context.Context) (*corev1.ConfigMap, error) {
 
        cm, err := r.KubeClient.CoreV1().
                ConfigMaps(r.DataPlaneConfigMapNamespace).
                Get(ctx, r.DataPlaneConfigMapName, metav1.GetOptions{})
 
        if apierrors.IsNotFound(err) {
-               cm, err = CreateDataPlaneConfigMap(ctx, r.KubeClient, r.DataPlaneConfigMapNamespace, r.DataPlaneConfigMapName, options...)
+               cm, err = r.createDataPlaneConfigMap(ctx)
+       }
+
+       if r.DataPlaneConfigMapTransformer != nil {
+               r.DataPlaneConfigMapTransformer(cm)
        }
 
        return cm, err
 }
 
-func CreateDataPlaneConfigMap(ctx context.Context, kubeClient kubernetes.Interface, namespace string, name string, options ...ConfigMapOption) (*corev1.ConfigMap, error) {
+func (r *Reconciler) createDataPlaneConfigMap(ctx context.Context) (*corev1.ConfigMap, error) {
        cm := &corev1.ConfigMap{
                TypeMeta: metav1.TypeMeta{},
                ObjectMeta: metav1.ObjectMeta{
-                       Name:      name,
-                       Namespace: namespace,
+                       Name:      r.DataPlaneConfigMapName,
+                       Namespace: r.DataPlaneConfigMapNamespace,
                },
                BinaryData: map[string][]byte{
                        ConfigMapDataKey: []byte(""),
                },
        }
 
-       for _, opt := range options {
-               opt(cm)
-       }
-
-       return kubeClient.CoreV1().ConfigMaps(namespace).Create(ctx, cm, metav1.CreateOptions{})
+       return r.KubeClient.CoreV1().ConfigMaps(r.DataPlaneConfigMapNamespace).Create(ctx, cm, metav1.CreateOptions{})
 }
 
 // GetDataPlaneConfigMapData extracts contract from the given config map.
diff --git a/control-plane/pkg/reconciler/broker/broker.go b/control-plane/pkg/reconciler/broker/broker.go
index 230ae987..922c9f84 100644
--- a/control-plane/pkg/reconciler/broker/broker.go
+++ b/control-plane/pkg/reconciler/broker/broker.go
@@ -69,8 +69,6 @@ type Reconciler struct {
        BootstrapServers string
 
        Prober prober.Prober
-
-       DataPlaneConfigMapLabeler base.ConfigMapOption
 }
 
 func (r *Reconciler) ReconcileKind(ctx context.Context, broker *eventing.Broker) reconciler.Event {
@@ -136,7 +134,7 @@ func (r *Reconciler) DoReconcileKind(ctx context.Context, broker *eventing.Broke
        }
 
        // Get contract config map.
-       contractConfigMap, err := r.GetOrCreateDataPlaneConfigMap(ctx, r.DataPlaneConfigMapLabeler)
+       contractConfigMap, err := r.GetOrCreateDataPlaneConfigMap(ctx)
        if err != nil {
                return statusConditionManager.FailedToGetConfigMap(err)
        }
@@ -284,7 +282,7 @@ func (r *Reconciler) DoFinalizeKind(ctx context.Context, broker *eventing.Broker
        logger := kafkalogging.CreateFinalizeMethodLogger(ctx, broker)
 
        // Get contract config map.
-       contractConfigMap, err := r.GetOrCreateDataPlaneConfigMap(ctx, r.DataPlaneConfigMapLabeler)
+       contractConfigMap, err := r.GetOrCreateDataPlaneConfigMap(ctx)
        if err != nil {
                return fmt.Errorf("failed to get contract config map %s: %w", r.DataPlaneConfigMapAsString(), err)
        }
diff --git a/control-plane/pkg/reconciler/broker/namespaced_broker.go b/control-plane/pkg/reconciler/broker/namespaced_broker.go
index 1965ca00..5729afe1 100644
--- a/control-plane/pkg/reconciler/broker/namespaced_broker.go
+++ b/control-plane/pkg/reconciler/broker/namespaced_broker.go
@@ -24,20 +24,20 @@ import (
        appsv1 "k8s.io/api/apps/v1"
        corev1 "k8s.io/api/core/v1"
        "k8s.io/apimachinery/pkg/api/errors"
-       apierrors "k8s.io/apimachinery/pkg/api/errors"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
        "k8s.io/apimachinery/pkg/types"
        "k8s.io/client-go/kubernetes/scheme"
        corelisters "k8s.io/client-go/listers/core/v1"
        "k8s.io/utils/pointer"
+       eventing "knative.dev/eventing/pkg/apis/eventing/v1"
+       "knative.dev/pkg/reconciler"
+       "knative.dev/pkg/resolver"
+
        "knative.dev/eventing-kafka-broker/control-plane/pkg/config"
        "knative.dev/eventing-kafka-broker/control-plane/pkg/kafka"
        "knative.dev/eventing-kafka-broker/control-plane/pkg/prober"
        "knative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/base"
-       eventing "knative.dev/eventing/pkg/apis/eventing/v1"
-       "knative.dev/pkg/reconciler"
-       "knative.dev/pkg/resolver"
 )
 
 // TODO: 1.
@@ -92,11 +92,6 @@ func (r *NamespacedReconciler) ReconcileKind(ctx context.Context, broker *eventi
 
        br := r.createReconcilerForBrokerInstance(broker)
 
-       err := r.reconcileDataPlaneConfigMap(ctx, broker)
-       if err != nil {
-               return fmt.Errorf("failed to create/update data plane config map: %s, error: %v", fmt.Sprintf("%s/%s", broker.Namespace, r.Env.DataPlaneConfigMapName), err)
-       }
-
        manifest, err := r.getManifest(broker)
        if err != nil {
                return fmt.Errorf("unable to transform dataplane manifest. namespace: %s, broker: %v, error: %v", broker.Namespace, broker, err)
@@ -134,6 +129,11 @@ func (r *NamespacedReconciler) createReconcilerForBrokerInstance(broker *eventin
 
                        DataPlaneNamespace:          broker.Namespace,
                        DataPlaneConfigMapNamespace: broker.Namespace,
+
+                       DataPlaneConfigMapTransformer: func(cm *corev1.ConfigMap) {
+                               kafka.NamespacedDataplaneLabelConfigmapOption(cm)
+                               appendBrokerAsOwnerRef(broker)
+                       },
                },
                Env:                        r.Env,
                Resolver:                   r.Resolver,
@@ -141,8 +141,6 @@ func (r *NamespacedReconciler) createReconcilerForBrokerInstance(broker *eventin
                NewKafkaClusterAdminClient: r.NewKafkaClusterAdminClient,
                BootstrapServers:           r.BootstrapServers,
                Prober:                     r.Prober,
-
-               DataPlaneConfigMapLabeler: kafka.NamespacedDataplaneLabelConfigmapOption,
        }
 }
 
@@ -236,30 +234,3 @@ func setImagesForDeployments(imageMap map[string]string) mf.Transformer {
                return scheme.Scheme.Convert(deployment, resource, nil)
        }
 }
-
-func (r *NamespacedReconciler) reconcileDataPlaneConfigMap(ctx context.Context, broker *eventing.Broker) error {
-       cm, err := r.KubeClient.CoreV1().
-               ConfigMaps(broker.Namespace).
-               Get(ctx, r.Env.DataPlaneConfigMapName, metav1.GetOptions{})
-
-       if err != nil && !apierrors.IsNotFound(err) {
-               return err
-       }
-
-       if apierrors.IsNotFound(err) {
-               // not found, create
-               _, err = base.CreateDataPlaneConfigMap(ctx, r.KubeClient, broker.Namespace, r.Reconciler.DataPlaneConfigMapName, appendBrokerAsOwnerRef(broker), kafka.NamespacedDataplaneLabelConfigmapOption)
-               return err
-       }
-
-       // found, update
-
-       if refs, appended := appendOwnerRef(cm.GetOwnerReferences(), broker); appended {
-               cm.SetOwnerReferences(refs)
-               // only update if we actually appended an ownerRef
-               _, err = r.KubeClient.CoreV1().ConfigMaps(broker.Namespace).Update(ctx, cm, metav1.UpdateOptions{})
-               return err
-       }
-
-       return nil
-}
```

Reply:
with this we also avoid unnecessary API server calls in `reconcileDataPlaneConfigMap`

Reply:
> This additional parameter should not be necessary.

Well, it was impossible for me to get all of these from 1 generic sentence.

I wrote that `reconcileDataPlaneConfigMap` because the existing `GetOrCreateDataPlaneConfigMap` didn't work like I wanted. Specifically, it was checking for existence and if CM was not there it was creating it. However, it is not updating the CM when it already existed.

I guess this approach would work. Except, it won't update the CM immediately and update it later with a label and owner. I need to think more about this.

Reply:
> I guess this approach would work. Except, it won't update the CM immediately and update it later with a label and owner. I need to think more about this.

No, there are some cases we need to think about...

It is safer to apply the label and the owner immediately. Otherwise, if there's something wrong until the rest of the reconciliation, we could lose the label and the ownerRef information until the update is happening.

I am talking about the case that the CM already exists. In that case, the label would be there anyway and that's good, but we might lose the ownerRef in case of a reconciliation problem.

What do you think?

Reply:
We can still call update here [1] but it's not necessary, however, at least the logic is encapsulated in one single place without duplication

[1]

```
+       if r.DataPlaneConfigMapTransformer != nil {
+               r.DataPlaneConfigMapTransformer(cm)
+               // update
        }
```

Reply:
> I am talking about the case that the CM already exists. In that case, the label would be there anyway and that's good, but we might lose the ownerRef in case of a reconciliation problem.

There shouldn't be a problem with this since, in the worst case, the ConfigMap will be deleted and then recreated when the Broker is reconciled again.